design/proposals/variable_provision.md

```diff
@@ -0,0 +1,295 @@
+  ...
+  ```
+
+- The owner of the value uses the Conjur API or CLI to store the value in the Variable:
+  
+  ```sh-session
+  $ conjur variable values add my-api-key abcdef12345
+  ...
+  ```
+
+- Once stored, access to the variable may be granted to applications, and those
+  applications can use value using the Conjur API, Summon, Secretless, etc.
+
+## Issue description
+
+There are a couple of issues this solution intends to address:
+
+### 'Browsing-up` for secrets management
+
+"Browse-up" is defined as [[ref]](https://www.ncsc.gov.uk/whitepaper/security-architecture-anti-patterns#section_3):
+
+> When administration of a system is performed from a device which is less trusted
+> than the system being administered.
+
+Unless performed only on dedicated, controlled workspaces, the `variable values add`
+step in our existing workflow necessitates that all secrets stored in Conjur (a
+"trusted system") must first travel through a user's workstation (a "less trusted system").
+
+### Secrets Domain Knowledge Required
+
+This isn't an "issue" as much as an opportunity for added value. Conjur today serves
+essentially as a database with encryption and audit. While these are crucial, as
+a Secrets store and Access Control system, there is an opportunity to provide more
+access control domain knowledge in Conjur itself, rather than requiring this as
+pre-requisite knowledge a user needs.
+
+An example of this is X.509 Certificates used for authentication, secure communications,
+and data integrity. While it may be simple enough to understand why I would want
+to use Certificates in my applications and infrastructure. Creating and maintaining
+these certificates, and a public key infastructure (PKI) is
+[anything but trivial](https://smallstep.com/blog/everything-pki/). And mistakes
+here can invalidate the security gains from using certificates in the first place.
+
+### Human Intervention Required
+
+As application deployments scale up, the need to automate the creation and rotation
+of credentials will continue to grow as human provisioning of credentails becomes
+a bottleneck. Our API enables end-users to bolt this automation onto Conjur, but
+we have an opportunity to enable more just-in-time creation of these credentails
+on behalf the user to make this kind of automation more widely available.
+
+### Credential Provenance
+
+While the information in Conjur is useful for auditing the *use* of a credential,
+Conjur is missing any ability to audit the *provenance* of a credential or its current
+validity.
+
+## Solution
+
+The solution I propose to address these issues is **variable provisioning**, in which
+Conjur is configured (through the policy where a variable is defined) to provision
+the secret value for a variable when the policy for that variable is loaded.
+
+For example, defining a variable to store an RSA private key could be represented
+in policy as:
+
+```yaml
+- !variable
+  id: private-key
```

Comment:
How could we capture the public key in a variable as well?

Reply:
At the moment in prototype, the `rsa` provisioner is the equivalent of:
```
openssl genrsa ...
```

The resulting PEM actually contains both the private key and public key, so you could have an rsa public key provisioner that just extracts the public key portion if you needed it, similar to how we reference the private key in the `x509-certificate` provisioner.

I'm not aware of many (any?) use cases where you want the public key apart from a X.509 certificate that identifies it though...